### PR TITLE
Add pivot_wider function

### PR DESCRIPTION
**Pull request by drrajeshtalluri**
Added pivot_wider function which performs unstack similar to the dplyr function pivot_wider.

### CODE REVIEW

**Comment by xiaodaigh** [comment]
@drrajeshtalluri Are you able to add some test cases in `./tests`? As a minimum, include [the examples (or their equivalents) from the `tidyr::pivot_wider` doc](https://tidyr.tidyverse.org/reference/pivot_wider.html#examples)

**Comment by drrajeshtalluri** [comment]
Sure, I will write some tests. I am adding some of my own tests for now, as I could not get the data for those examples in the R vignette. They are not present in RDatasets as it does not contain tidyr datasets.  Also, in several places, Ddply.jl  is mentioned in the code. Probably reinitialize the package or something similar to remove them and avoid confusion.

**Comment by drrajeshtalluri** [comment]
I have put some tests in. I am new to package development and maintenance. Let me know if these look ok.

File: src/Pivot.jl [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+│ 2   │ 2     │ 5     │ 8     │ ce_val1_2 │ cf_val1_2 │ de_val1_2 │ df_val1_2 │ ce_val2_2 │ cf_val2_2 │ de_val2_2 │ df_val2_2 │
+│ 3   │ 3     │ 6     │ 9     │ ce_val1_3 │ cf_val1_3 │ de_val1_3 │ df_val1_3 │ ce_val2_3 │ cf_val2_3 │ de_val2_3 │ df_val2_3 │
+"""
+function pivot_wider(df::AbstractDataFrame; names_from = nothing, values_from = nothing)
+ if names_from isa Symbol
+    names_from = [names_from]
+ end
+ if values_from isa Symbol
+    values_from = [values_from]
+ end
```
Comment (xiaodaigh) [inline src/Pivot.jl]: indentation is off. Also, given that dataframes can handle symbols and string for column names. Both should be handled here.

File: src/Pivot.jl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    id_cols = setdiff(Symbol.(names(df)), names_from, values_from)
+    dfa = select(df, :)
```
Comment (xiaodaigh) [inline src/Pivot.jl]: This makes a copy of `df` which can cause issues for large datasets due to excess RAM usage. Can the algorithm be tweek to not make a copy of `df`?

File: src/Pivot.jl [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    dfa[!,:cname_temp] = [join(Array(r), "_") for r in eachrow(select(dfa, names_from))]
```
Comment (xiaodaigh) [inline src/Pivot.jl]: would appreciate some comments on what it's attempting to do as ppl can forget and it makes maintenance harder in the future.

File: src/Pivot.jl [hunk 4]
```diff
@@ -0,0 +1,71 @@
+"""
+unstack_name(df::AbstractDataFrame, rowkeys, colkey, value)
+
+A helper function for unstack to name the columns being unstacked using the column name the values are present in.
+"""
+function unstack_name(df::AbstractDataFrame, rowkeys, colkey, value)
+    unstack(df::AbstractDataFrame, rowkeys, colkey, value, renamecols = x -> Symbol(string(value), "_", x))
+end
```
Comment (xiaodaigh) [inline src/Pivot.jl]: Do we need this function? I see that it's only used once and is not exported. In that case, I would prefer the code to be in the in main pivot_wider function and just have brief comments describing what it does.

File: src/Pivot.jl [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+
+
+"""
+pivot_wider(df::AbstractDataFrame; names_from =  names_cols, values_from = values_cols)
+
+A function to widen a data frame (e.g. unstack)  but in the frameowrk of dplyr.
```
Comment (xiaodaigh) [inline src/Pivot.jl]: Add a link to the tidyr documentation for pivot_wider?

**Review by xiaodaigh — CHANGES_REQUESTED** [review]

File: src/Pivot.jl (hunk 1 above)
Comment (drrajeshtalluri) [reply src/Pivot.jl]: Ok Done!

File: src/Pivot.jl (hunk 2 above)
Comment (drrajeshtalluri) [reply src/Pivot.jl]: I removed the copying by adding and removing columns from the original dataframe.

File: src/Pivot.jl (hunk 3 above)
Comment (drrajeshtalluri) [reply src/Pivot.jl]: Added some comments to explain the code.

File: src/Pivot.jl (hunk 5 above)
Comment (drrajeshtalluri) [reply src/Pivot.jl]: Added the link to dplyr pivot_wider page.

File: src/Pivot.jl (hunk 4 above)
Comment (drrajeshtalluri) [reply src/Pivot.jl]: I actually needed to broadcast this by accessing the value columns for the `renamecols` argument. However, I did not have enough knowledge to create a list of anonymous functions to broadcast here.  This should be probably simple without defining a function. I have now written an inline function inside the main function. I thought you said that it was gonna have performance issues, so I made it a separate function. So, please let me know if you know a way to do the broadcast without defining a function.

File: src/Pivot.jl [hunk 6]
```diff
@@ -0,0 +1,87 @@
+"""
+pivot_wider(df::AbstractDataFrame; names_from =  names_cols, values_from = values_cols)
```
Comment (xiaodaigh) [inline src/Pivot.jl]: The argument list is very shortened from `tidyr::pivot_wider`. I think it's ok to have fewer features than pivot_wider.

I wonder if there's a way to let user-know that not all options are implemented.  I guess the docstring is fine.

File: src/Pivot.jl [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+A function to widen a data frame (e.g. unstack)  but in the frameowrk of dplyr.
```
Comment (xiaodaigh) [inline src/Pivot.jl]:
```suggestion
A function to widen a data frame (i.e. unstack)  but in the framework of dplyr.
```

File: src/Pivot.jl [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+   ## Defining id_cols as all columns other than names and values columns
+   id_cols = setdiff(Symbol.(names(df)), names_from, values_from)
+   if names_from isa Array{String}
+      id_cols = setdiff(names(df), names_from, values_from)
+   end
+   
+   # Concatenating multiple name columns with "_" and creating a single name column cname_temp
+   # This may cause a bug if someone names their column cname_temp (Dont know how to work around this issue: creating a temporary column using it and then deleting it without conflicts)
+   df[!,:cname_temp] = [join(Array(r), "_") for r in eachrow(select(df, names_from))]
```
Comment (xiaodaigh) [inline src/Pivot.jl]: There is a `gen_sym` function. This name is too easy to clash.

File: src/Pivot.jl [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+   
+   # function to  unstack with column names added
+   unstack_name(df::AbstractDataFrame, rowkeys, colkey, value) = unstack(df::AbstractDataFrame, rowkeys, colkey, value, renamecols = x -> Symbol(string(value), "_", x))
+
+   # Broadcast unstacking on all value columns 
+   wide_dfs = unstack_name.(Ref(df), Ref(id_cols), Ref(:cname_temp), values_from)
```
Comment (xiaodaigh) [inline src/Pivot.jl]: In that case, maybe just list comprehension is sufficient

```suggestion
   wide_dfs = [unstack(df, id_cols, :cname_temp, value, renamecols = x -> string(value)*"_"*x)
            for value in values_from]
```

**Comment by xiaodaigh** [comment]
@drrajeshtalluri I notice a lot of indentations are off. Can I suggest you use something like JuliaFormatter.jl to format the code.

File: src/Pivot.jl [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+```
+3×11 DataFrame
+│ Row │ x     │ a     │ b     │ val1_c_e  │ val1_c_f  │ val1_d_e  │ val1_d_f  │ val2_c_e  │ val2_c_f  │ val2_d_e  │ val2_d_f  │
+│     │ Int64 │ Int64 │ Int64 │ String?   │ String?   │ String?   │ String?   │ String?   │ String?   │ String?   │ String?   │
+├─────┼───────┼───────┼───────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┤
+│ 1   │ 1     │ 4     │ 7     │ ce_val1_1 │ cf_val1_1 │ de_val1_1 │ df_val1_1 │ ce_val2_1 │ cf_val2_1 │ de_val2_1 │ df_val2_1 │
+│ 2   │ 2     │ 5     │ 8     │ ce_val1_2 │ cf_val1_2 │ de_val1_2 │ df_val1_2 │ ce_val2_2 │ cf_val2_2 │ de_val2_2 │ df_val2_2 │
+│ 3   │ 3     │ 6     │ 9     │ ce_val1_3 │ cf_val1_3 │ de_val1_3 │ df_val1_3 │ ce_val2_3 │ cf_val2_3 │ de_val2_3 │ df_val2_3 │
+"""
+function pivot_wider(df::AbstractDataFrame; names_from = nothing, values_from = nothing)
```
Comment (xiaodaigh) [inline src/Pivot.jl]: `pivot_wider(df)` will cause an error currently. If we allow `nothing` as the default then we might as error throw an error message as default.

```suggestion
function pivot_wider(df::AbstractDataFrame; names_from = error("Hadleyverse.pivot_wider: you must specify `names_from`"), values_from = error("Hadleyverse.pivot_wider: you must specify `values_from`"))
```

File: src/Pivot.jl [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+│ 6   │ 3     │ 6     │ 9     │ cf_val1_3 │ cf_val2_3 │ c      │ f      │
+│ 7   │ 1     │ 4     │ 7     │ de_val1_1 │ de_val2_1 │ d      │ e      │
+│ 8   │ 1     │ 4     │ 7     │ df_val1_1 │ df_val2_1 │ d      │ f      │
+│ 9   │ 2     │ 5     │ 8     │ de_val1_2 │ de_val2_2 │ d      │ e      │
+│ 10  │ 2     │ 5     │ 8     │ df_val1_2 │ df_val2_2 │ d      │ f      │
+│ 11  │ 3     │ 6     │ 9     │ de_val1_3 │ de_val2_3 │ d      │ e      │
+│ 12  │ 3     │ 6     │ 9     │ df_val1_3 │ df_val2_3 │ d      │ f      │
+
+```
+pivot_wider(df, names_from = [:cname1,:cname2], values_from = [:val1,:val2])
```
Comment (xiaodaigh) [inline src/Pivot.jl]:
```suggestion
pivot_wider(df; names_from = [:cname1,:cname2], values_from = [:val1,:val2])
```

File: src/Pivot.jl [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+    # remove created column cname_temp
+    select!(df, Not(tmp_name))
+
+    # check if there are multiple dataframes
+    if length(wide_dfs) >= 2
+        # remove all id columns for any other frame other than the first
+        select!.(@view(wide_dfs[2:end]), Ref(Not(id_cols)))
+        # concatenate all the dataframes horizontally
+        reduce((df1, df2) -> hcat(df1, df2, makeunique = true), wide_dfs)
```
Comment (xiaodaigh) [inline src/Pivot.jl]:
```suggestion
        hcat(wide_dfs..., makeunique = true)
```

**Review by xiaodaigh — CHANGES_REQUESTED** [review]

**Review by xiaodaigh — APPROVED** [review]
Thanks @drrajeshtalluri.

I just tidied up the code a bit!  Hope that was an enjoyable experience with you.

**Comment by xiaodaigh** [comment]
@drrajeshtalluri if you wish to implement the rest of the parameters in `pivot_wider`. More than happy to review the PR 👍 

image

**Comment by drrajeshtalluri** [comment]
> @drrajeshtalluri if you wish to implement the rest of the parameters in `pivot_wider`. More than happy to review the PR 👍
> 
> image

Thanks, I will keep working on it and add all the features from dplyr to make it complete. Will be a good learning experience for me.

**Pull request merged**